### ISSUE TITLE
LIMIT doesn't work with custom read cypher query

### ISSUE THREAD

**Issue by moxious** [neo4j-contrib/neo4j-spark-connector#198]
See screenshot.

If you take out the LIMIT 10 this does work.  With the LIMIT 10 it's an error.

<img width="1222" alt="Screen Shot 2020-09-28 at 7 59 52 AM" src="https://user-images.githubusercontent.com/5471710/94429675-aadc8900-0160-11eb-950a-1388c49a83d5.png">

**Comment by utnaf** [comment]
@moxious can you try to put the limit before the return? This was it should work

`MATCH (h:Household) LIMIT 2 RETURN id(h) AS id`

Issue is that we already add a limit for the schema sampling and that way the query will result in:

```
MATCH (h:Household) RETURN id(h) AS id LIMIT 2
ORDER BY rand()
LIMIT 10
```

We can discuss the solutions:
- specify to put the limit before the return statement
- check if the limit is after the return, and remove it for the schema generation
- add an option for limiting the query result (useful for other type of reading as well)

**Comment by moxious** [comment]
We'll talk more about this.  I think I like your 3rd option the best.  The issue is that I don't want to restrict the cypher language any more than absolutely necessary in order to make things work.  Users would clearly expect that simple limits work in any order that is cypher valid, so we will need to brainstorm a bit how to make this possible, or if it's not practical, surface an error that lets the user fix it easily.

**Comment by moxious** [comment]
Another option:  detect cypher queries that use LIMIT / SKIP and reject query.   Do your limits & skips with Spark.

**Comment by jexp** [comment]
I would not do that. It’s error prone and would disable a wide range of queries that regularly use skip limit as part of their structure.

**Comment by moxious** [comment]
@jexp points of discussion here:

* SKIP and LIMIT may be actively bad when you're running partitioned spark queries.  How does this operate when you're doing it 4x in parallel?  Which records are being skipped in which partition?
* SKIP and LIMIT are already in use by the spark connector to implement parallelized reads.  So if your read query includes SKIP/LIMIT, then this clashes with the method we use to parallelize the reads.  The alternative might be, "You can use SKIP/LIMIT, but only if partitions=1, never otherwise"
* What SKIP & LIMIT are accomplishing is already trivially implementable with the Spark API when you get the data back, and so doing it at the cypher layer isn't saving anything.

Do we have cases where SKIP/LIMIT *must* be at the end of a Cypher query to return proper data?    Note that internal limits (within subqueries) will be allowed.

**Comment by conker84** [comment]
> @moxious can you try to put the limit before the return? This was it should work
> 
> `MATCH (h:Household) LIMIT 2 RETURN id(h) AS id`
> 
> Issue is that we already add a limit for the schema sampling and that way the query will result in:
> 
> ```
> MATCH (h:Household) RETURN id(h) AS id LIMIT 2
> ORDER BY rand()
> LIMIT 10
> ```
> 
> We can discuss the solutions:
> 
> * specify to put the limit before the return statement
> * check if the limit is after the return, and remove it for the schema generation
> * add an option for limiting the query result (useful for other type of reading as well)

I would go with the third option:

`option("query.limit", 100)`

**Comment by moxious** [comment]
Scenario:  user wants to fetch first 1 million of a dataset of 10 million nodes; in this case it would be important to push the limit into the cypher query, rather than just using .limit() on a 10 million dataset.

Currently thinking maybe `.option("query.limit", 100)` is the best approach?

**Comment by utnaf** [comment]
Agree with the `query.limit` option, I think it's more straightforward.

On the other side, do we have to explicitly tell the user to not SKIP/LIMIT their query if partitions > 1? Or we just want to tell them to avoid a final SKIP/LIMIT?

**Comment by utnaf** [comment]
So apparently we are already handling `query.limit` with the `query.count` option.
This option can be both an actual cypher count query or a number. The number works exactly like `query.limit`. 

I would leave it like this, not implement `query.limit` and document the `query.count` better.

Thoughts? @moxious @conker84

**Comment by moxious** [comment]
Using an explicit option is OK as an approach.  I don't love it, but there isn't a straightforward way to allow a final LIMIT/SKIP.

There's no need to duplicate an option, so yes please to improving the documentation on query.count.

The last part related to this issue is that ideally we'd have a better call-out error message with the connector so that if a person DOES use limit/skip in a cypher query, they shouldn't get a syntax error or a "please provide valid read query" they should get a message telling them not to do that, if possible.  That's a nice to have though, not a requirement of the ticket.   If you *can't* get a clear error message in that case (and you might not be able to since parsing Cypher is not a great approach) -- then add a FAQ.    "Why am I getting <error message here>?"   Don't use SKIP/LIMIT, etc.   Make sure to point out that *internal SKIP/LIMIT* in the middle of the query is still ok.

Add a unit test or two around this case, it's bound to trip some peopel up
